From 5d55db32d451f2378f901d9cd85adccfc570c08f Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:42 -0400
Subject: [fs] nfsd: Ensure that nfs4_open_delegation() references the delegation stateid

Message-id: <1410898849-22126-99-git-send-email-bfields@redhat.com>
Patchwork-id: 95854
O-Subject: [Patch RHEL7 098/165] nfsd: Ensure that nfs4_open_delegation() references the delegation stateid
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Ensure that nfs4_open_delegation() keeps a reference to the delegation
stateid until it is done working with it. Necessary step toward
client_mutex removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 67cb1279be27345ba6855c1aab9e90ef5f5ac645
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 4f19145..2df6af9 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -674,6 +674,7 @@ hash_delegation_locked(struct nfs4_delegation *dp, struct nfs4_file *fp)
  lockdep_assert_held(&state_lock);
  lockdep_assert_held(&fp->fi_lock);
 
+ atomic_inc(&dp->dl_stid.sc_count);
  dp->dl_stid.sc_type = NFS4_DELEG_STID;
  list_add(&dp->dl_perfile, &fp->fi_delegations);
  list_add(&dp->dl_perclnt, &dp->dl_stid.sc_client->cl_delegations);
@@ -3704,6 +3705,7 @@ nfs4_open_delegation(struct svc_fh *fh, struct nfsd4_open *open,
  dprintk("NFSD: delegation stateid=" STATEID_FMT "\n",
   STATEID_VAL(&dp->dl_stid.sc_stateid));
  open->op_delegate_type = NFS4_OPEN_DELEGATE_READ;
+ nfs4_put_stid(&dp->dl_stid);
  return;
 out_no_deleg:
  open->op_delegate_type = NFS4_OPEN_DELEGATE_NONE;
-- 
1.7.1